src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -201,6 +211,12 @@ internal Task Ensure(int count, bool async, bool readingNotifications)
                         // We can't stop in a finally block because Connector.Break() will dispose the buffer and the contained
                         // _timeoutCts
                         buffer.Cts.Stop();
+                        if (!async)
+                        {
+                            syncCancellationRegistration.Dispose();
+                            if (finalCt.IsCancellationRequested)
+                                connector.TryThrowBreakReason();
```

Comment:
We can put a Debug.Fail inside the condition after this line - if IsCancellationRequested is true, we should always throw here.

Reply:
Not really - we might have failed due to IO error. Also, we might dispose the registration first, and only after that the cancellation was requested, so the connection is not broken.